### ISSUE TITLE
Admin theme "Modules"?

### ISSUE THREAD

**Issue by xgp** [keycloak/keycloak-admin-ui#1219]
I've recently encountered a problem that I thought it might be constructive to raise here. 

It seems to me that there are (at least) two different kinds of admin themes. 
1. change the look-and-feel/style of the default Admin UI.
2. necessary to administer extension functionality. E.g. in order to provide a UI for admins to use [this SCIM extension](https://github.com/Captain-P-Goldfish/scim-for-keycloak), the developer has created a custom theme.

The problem I've encountered is that one may want to combine multiple themes. This came up when we were adding the theme for the SCIM extension in the example above, already had an internal theme that exposed UI for our extensions to the event config, and had customized the style of the Admin UI in another theme. Our initial strategy had been to have our style theme use our event theme as a parent. This is impossible for more than one theme, so we had to do some messy and direct integration of the 3rd party theme in order to make it work as desired.

The question I'd like to pose here is if there is a way to define "Modules" for extension to the Admin UI that would enable us to install multiple admin "themes". It would need to include some kind of way for identifying and/or resolving conflicts between extensions. I don't have a concrete proposal of how it would work, but I wanted to raise it here in case it's something others have thought of and want to resolve.

**Comment by edewit** [comment]
Well currently you can [inherit](https://github.com/Captain-P-Goldfish/scim-for-keycloak/blob/5987a78c6ebf56cabd7d3343f8fffdb24005aef7/scim-for-keycloak-server/src/main/resources/theme/scim/admin/theme.properties#L1) from a theme and we have more integration points planned we might have reusable parts of the console at some point

**Comment by xgp** [comment]
Yes, what I'm suggesting would be something like multiple inheritance. If you have multiple admin themes that are built to administer extension functionality, it would be useful to have a mechanism to include more than one. Are the "more integration points" you mentioned documented somewhere? Or is there a spec for the new admin console?

**Comment by edewit** [comment]
At the moment we don't have concrete plans on this yet. Here you can find the [design docs](https://www.keycloak.org/keycloak-community/design/admin-console/#/)

**Comment by xgp** [comment]
Hi @edewit . Checking in on this enhancement now that the new admin console is default. I know a few Keycloak extension authors that are eager to port the admin portion of their extensions to the new console.

Per your comment:
> we have more integration points planned

Can you share if there has been any more discussion on this topic, or if you have come to a decision about how plugins/extensions to the admin console might be achieved?

Thank you! 

(btw, the new admin console looks great!)

**Comment by edewit** [comment]
In the end we didn't build anything to help you with extending the UI. The path for extending it is fork our repo and then build your stuff against it. Then if you use quarkus you can build an `jar` en move it into you distribution.

**Comment by xgp** [comment]
@edewit Thanks for the response. That's too bad, as I know there are extension authors that would like to port extensions to the new console. cc @sventorben @Captain-P-Goldfish

Is there an appetite for accepting a contribution in this regard? One of the people I work with has an idea of how to create a simple plugin interface, such that extensions could be added as jars without forking the repo.

In any case, is the current admin UI now considered stable? We have been trying to maintain an extension for the last year in the manner you have described, but every release seems to come with a massive refactoring of the codebase, making it very labor intensive to port to each new version.

**Comment by Captain-P-Goldfish** [comment]
I lso asked for this already here: https://github.com/keycloak/keycloak/discussions/14327#discussioncomment-3621008

**Comment by jonkoops** [comment]
> Is there an appetite for accepting a contribution in this regard?

Absolutely. If there are idea's on what kind of interface plugins would need to work then we are totally open for suggestions.

**Comment by xgp** [comment]
This is not a proposal, but I wanted to add some notes here, in case it is useful.

Most Keycloak admin extensions need:

- Access to the left nav to add a menu item.
- Access to the token and userinfo of the logged in user.
- Access to run admin API queries. Although this can be achieved with just the token + the stock admin API lib, there may be some reusable components that would make this easier.
- Access to the router?

Is it sufficient to have a dedicated "Extensions" section of the left nav that you can "+" add things to? This could pull from an SPI registry. Need an additional endpoint to show what is available, and put "registered" extensions in realm attributes. Each "extension" (= implementation of the new SPI) can be a set of Resources and static assets.

How do these "pages" get rendered? Does this force an assumption that all extensions must maintain compatibility with the UI framework(s) used and the extensive refactoring always underway? Is there a way to isolate the core from the extension code so they don't conflict? Namespacing? Is it possible to give the extension a `div` and let it do whatever it wants?

Here's an example diff of keycloak-ui for an extension ([keycloak-orgs](https://github.com/p2-inc/keycloak-orgs)) where I tried to change as little as possible in the parent keycloak-ui project, while adding my extension to the nav and importing minimal utils from the parent project. This may give some inspiration for the minimum required extension points: https://github.com/keycloak/keycloak-ui/compare/19.0.3...p2-inc:keycloak-ui:19.0.3_orgs

I found these to be an interesting/relevant read:
- https://www.pagerduty.com/eng/react-embedded-apps/
- https://micro-frontends.org/
And  [single-spa](https://single-spa.js.org/) to be a project that we might want to explore, as it could allow those that have invested heavily in their angular1 admin extensions to use that without having to build it in react from scratch.

**Comment by edewit** [comment]
@xgp we had something like this in the account console, not only did it let people add "pages" it also let people override existing pages. But there is a heavy downside to this, it meant that loading times increased and adding pages was still hard so we are wondering if it isn't easier to "just" clone and use your approach where you add something without changing the existing code to much

**Comment by xgp** [comment]
Hi @edewit, thanks for taking a look at this. Definitely agree that priorities should be to keep page load time the same and make new extension development fairly easy. I think that there is an approach that would lazy-load the code for each extension. Let me put together a quick prototype of what I'm talking about so it's not abstract.

**Comment by BradNut** [comment]
@xgp Do you know if there was an update on this?

**Comment by xgp** [comment]
Hi @BradNut . We haven't dedicated time to this. We have a setup that works for us, but it is not completely "modular", as we would like. And frankly, without guidance from the Keycloak team, we aren't going to take the time to formalize something that they will likely reject.
Our approach is to have a simple SPI that is implemented by extension authors that gives them 1) a tab in the left nav, 2) a mechanism to serve resources (like a theme) and 3) a div in the main content area when the user clicks on the tab. There are some other hacks to give their code access to the token and such, but that's basically it. We wanted to keep it as simple as possible, and avoided any deep integration with the current codebase, as we've found it changes fairly rapidly. There is actually not a hard dependency on running the extensions inside of the current admin UI.
